components/camel-kubernetes/src/main/java/org/apache/camel/component/kubernetes/customresources/KubernetesCustomResourcesProducer.java

```diff
@@ -0,0 +1,245 @@
+        }
+
+        switch (operation) {
+
+            case KubernetesOperations.LIST_CUSTOMRESOURCES:
+                doList(exchange, operation);
+                break;
+
+            case KubernetesOperations.LIST_CUSTOMRESOURCES_BY_LABELS_OPERATION:
+                doListByLabels(exchange, operation);
+                break;
+
+            case KubernetesOperations.GET_CUSTOMRESOURCE:
+                doGet(exchange, operation);
+                break;
+
+            case KubernetesOperations.DELETE_CUSTOMRESOURCE:
+                doDelete(exchange, operation);
+                break;
+
+            case KubernetesOperations.CREATE_CUSTOMRESOURCE:
+                doCreate(exchange, operation);
+                break;
+
+            default:
+                throw new IllegalArgumentException("Unsupported operation " + operation);
+        }
+    }
+
+    protected void doList(Exchange exchange, String operation) throws Exception {
+        String namespaceName = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_NAMESPACE_NAME, String.class);
+        JsonObject customResourcesListJSON = new JsonObject(
+                getEndpoint().getKubernetesClient().customResource(getCRDContext(exchange.getIn())).list(namespaceName));
+        LOG.info(customResourcesListJSON.toString());
+
+        JsonArray customResourcesListItems;
+        if (customResourcesListJSON.getCollection("items") != null) {
+            customResourcesListItems = new JsonArray(customResourcesListJSON.getCollection("items"));
+        } else {
+            customResourcesListItems = new JsonArray();
+        }
+
+        MessageHelper.copyHeaders(exchange.getIn(), exchange.getOut(), true);
+        exchange.getOut().setBody(customResourcesListItems);
+    }
+
+    protected void doListByLabels(Exchange exchange, String operation) throws Exception {
+        String namespaceName = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_NAMESPACE_NAME, String.class);
+        Map<String, String> labels = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_CRD_LABELS, Map.class);
+        JsonObject customResourcesListJSON = new JsonObject(
+                getEndpoint().getKubernetesClient().customResource(getCRDContext(exchange.getIn())).list(namespaceName));
+        LOG.info(customResourcesListJSON.toString());
+        JsonArray customResourcesListItems = new JsonArray(customResourcesListJSON.getCollection("items"));
+
+        MessageHelper.copyHeaders(exchange.getIn(), exchange.getOut(), true);
+        exchange.getOut().setBody(customResourcesListItems);
+    }
+
+    protected void doGet(Exchange exchange, String operation) throws Exception {
+        String customResourceName = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_CRD_INSTANCE_NAME, String.class);
+        String namespaceName = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_NAMESPACE_NAME, String.class);
```

Comment:
Since you are already obtaining the namespace [here](https://github.com/apache/camel/pull/4541/files#diff-6ca11b2d4ce0280f5835a826f21758273973e7cbf43218bdcf87740b3c079990R60, would it be more efficient to just pass the obtained namespace to the function instead of re-obtaining the namespace again?

Reply:
OK.  I make changes.  Thanks!